using_restricted_array.rb

```diff
@@ -6,44 +6,115 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
+
 def length(array)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] == nil
+    index += 1
+  end
+  return index
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] == nil
+    print "#{array[index]} "
+    index += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  index = 0
+  until array[index] > length
+    if array[index] == value_to_find
+      return true
+    else
+      return false
+      index =+ 1
+    end
+  end
 end
 
+
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
+
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  # puts "NOT IMPLEMENTED"
+  a = 0
+  while a < length
+    b = 0
+    while b < length-a-1
+      if array[b] > array[b+1] # swap
+        temp = array[b]
+        array[b] = array[b+1]
+        array[b+1] = temp
+      end
+      b += 1
+    end
+    a += 1
+  end
+  return array[length-1]
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
+
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  # puts "not implemented"
+  min = array[0]
+  index = 1
+  while index < length
+    min = array[index] if array[index] < min
+    index += 1
+  end
+  return min
```

Comment:
This is an O(n) algorithm. Can you make your find_largest do something similar to this?